sse2neon.h

```diff
@@ -1405,38 +1345,316 @@ FORCE_INLINE __m128i _mm_srai_epi16(__m128i a, int count)
+    int32x4_t b = vreinterpretq_s32_m128i(_b);
+    // NEON and SSE both use zero on > 31, but NEON can't be
+    // negative or it will shift right.
+    int32x4_t mask = vandq_s32(b, vdupq_n_s32(0x7fffffff));
+    return vreinterpretq_m128i_u32(vshlq_u32(a, mask));
+}
+
+
+FORCE_INLINE __m128i _mm_sllv_epi64(__m128i _a, __m128i _b)
+{
+    uint64x2_t a = vreinterpretq_u64_m128i(_a);
+    int64x2_t b = vreinterpretq_s64_m128i(_b);
+    // NEON and SSE both use zero on > 63, but NEON can't be
+    // negative or it will shift right.
+    int64x2_t mask = vandq_s64(b, vdupq_n_s64(0x7fffffffffffffff));
+    return vreinterpretq_m128i_u64(vshlq_u64(a, mask));
+}
+
+FORCE_INLINE __m128i _mm_sll_epi16(__m128i a, __m128i b)
+{
+    int16x8_t duped =
+       vdupq_lane_s16(
+          vget_low_s16(vreinterpretq_s16_m128i(b)), 0
+       );
+    return _mm_sllv_epi16(a, vreinterpretq_m128i_s16(duped));
+}
+
+FORCE_INLINE __m128i _mm_sll_epi32(__m128i a, __m128i b)
+{
+    int32x4_t duped =
+       vdupq_lane_s32(
+          vget_low_s32(vreinterpretq_s32_m128i(b)), 0
+       );
+    return _mm_sllv_epi32(a, vreinterpretq_m128i_s32(duped));
+}
+
+FORCE_INLINE __m128i _mm_sll_epi64(__m128i a, __m128i b)
+{
+    int64x2_t duped =
+       vdupq_lane_s64(
+          vget_low_s64(vreinterpretq_s64_m128i(b)), 0
+       );
+    return _mm_sllv_epi64(a, vreinterpretq_m128i_s64(duped));
+}
+
+FORCE_INLINE __m128i _mm_srlv_epi16(__m128i _a, __m128i _b)
+{
+    uint16x8_t a = vreinterpretq_u16_m128i(_a);
+    int16x8_t b = vreinterpretq_s16_m128i(_b);
+    // NEON and SSE both use zero on > 15, but NEON can't be
+    // negative or it will shift left.
+    int16x8_t mask = vandq_s16(b, vdupq_n_s16(0x7fff));
+    int16x8_t reversed = vnegq_s16(mask);
+    return vreinterpretq_m128i_u16(vshlq_u16(a, reversed));
+}
+
+FORCE_INLINE __m128i _mm_srlv_epi32(__m128i _a, __m128i _b)
+{
+    uint32x4_t a = vreinterpretq_u32_m128i(_a);
+    int32x4_t b = vreinterpretq_s32_m128i(_b);
+    // NEON and SSE both use zero on > 31, but NEON can't be
+    // negative or it will shift left.
+    int32x4_t mask = vandq_s32(b, vdupq_n_s32(0x7fffffff));
+    int32x4_t reversed = vnegq_s32(mask);
+    return vreinterpretq_m128i_u32(vshlq_u32(a, reversed));
+}
+
+FORCE_INLINE __m128i _mm_srlv_epi64(__m128i _a, __m128i _b)
+{
+    uint64x2_t a = vreinterpretq_u64_m128i(_a);
+    int64x2_t b = vreinterpretq_s64_m128i(_b);
+    // NEON and SSE both use zero on > 63, but NEON can't be
+    // negative or it will shift right.
+    int64x2_t mask = vandq_s64(b, vdupq_n_s64(0x7fffffffffffffff));
+    int64x2_t reversed = vsubq_s64(vdupq_n_s64(0), mask);
+    return vreinterpretq_m128i_u64(vshlq_u64(a, reversed));
+}
+
+
+FORCE_INLINE __m128i _mm_srl_epi16(__m128i a, __m128i b)
+{
+    int16x8_t duped =
+       vdupq_lane_s16(
+          vget_low_s16(vreinterpretq_s16_m128i(b)), 0
+       );
+    return _mm_srlv_epi16(a, vreinterpretq_m128i_s16(duped));
+}
```

Comment:
According to the [_mm_srl_epi16](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#expand=4961,4961,4955,1885,452,3864,6077,5152,5143,5214,5205,5143,5205,5205,5214,5196,5205,5214,3980,3533,2938,1615,6050,6077,5264,5476,781,2432,3145,5225,5420,5326,5538,5476,5391,5449,5476,5538,5476,5556,5547,5420,5294,5326,5264,5538,5476&text=_mm_srl_epi16), what will happen if **b[17]**, **b[16]** and **b[0]** are set to 1?
I think the corresponding result is 0 but your implementation would do shift right 1 on the original data?